src/transaction.ts

```diff
@@ -44,35 +62,47 @@ export function assembleTransaction(
     minAccountSequenceLedgerGap: raw.minAccountSequenceLedgerGap,
     extraSigners: raw.extraSigners,
   });
-  for (let i = 0; i < raw.operations.length; i++) {
-    const rawOp = raw.operations[i];
-    if ("function" in rawOp) {
-      const sim = simulated[i];
-      if (!sim) {
-        throw new Error("missing simulated operation");
-      }
-      let footprint = sim.footprint ?? rawOp.footprint;
-      if (!(footprint instanceof xdr.LedgerFootprint)) {
-        footprint = xdr.LedgerFootprint.fromXDR(footprint.toString(), "base64");
-      }
-      const auth = (sim.auth ?? rawOp.auth).map((a) =>
-        a instanceof xdr.ContractAuth
-          ? a
-          : xdr.ContractAuth.fromXDR(a.toString(), "base64"),
-      );
-      // TODO: Figure out a cleaner way to clone these operations
-      txn.addOperation(
-        Operation.invokeHostFunction({
-          function: rawOp.function,
-          parameters: rawOp.parameters,
-          footprint,
-          auth,
-        }),
-      );
-    } else {
-      // TODO: Handle this.
-      throw new Error("Unsupported operation type");
-    }
+
+  // apply the pre-built Auth from simulation onto each Tx/Op/HostFunction
+  // invocation
+  const authDecoratedHostFunctions = simulation.results.map(
+    (functionSimulationResult, i) => {
+      const hostFn: xdr.HostFunction = rawInvokeHostFunctionOp.functions[i];
+      hostFn.auth(buildContractAuth(functionSimulationResult.auth));
+      return hostFn;
+    },
+  );
+
+  txnBuilder.addOperation(
+    Operation.invokeHostFunctions({
+      functions: authDecoratedHostFunctions,
+    }),
+  );
+
+  // apply the pre-built Soroban Tx Ext from simulation onto the Tx
+  txnBuilder.setExt(buildExt(simulation.transactionData));
+
+  return txnBuilder.build();
+}
+
+function buildExt(sorobanTxDataStr: string) {
+  const sorobanTxData: xdr.SorobanTransactionData = xdr.SorobanTransactionData.fromXDR(
+    sorobanTxDataStr,
+    "base64",
+  );
+
+  // TODO - remove this workaround to use constructor on js-xdr union
+  //       and use the typescript generated static factory method once fixed
+  //       https://github.com/stellar/dts-xdr/issues/5
+  // @ts-ignore
+  const txExt = new xdr.TransactionExt(1, sorobanTxData);
+  return txExt;
+}
```

Comment:
related to my comment here: https://github.com/stellar/js-stellar-base/pull/604#discussion_r1194273887, this is what I meant about a more ergonomic API. Here, you would just do `txnBuilder.setSorobanData(simulation.transactionData)` and not worry about the extension/xdr/base64 nonsense.

Reply:
yes, I can refactor that, it would encapsulate the xdr.TransactionExt construction stuff to just one place in stellar-base tx-builder, if/when new TransactionExt union types later, the builder would need new method for each union type, but that is a fine trade-off for having the stronger types up front on the builder to get that more fluent client usage.

Reply:
refactored builder, using `setSorobanData()` - https://github.com/stellar/js-stellar-base/pull/604/commits/177b9e9acb19274b22192a9b0720c32099b19f1d

Reply:
> the builder would need new method for each union type, but that is a fine trade-off for having the stronger types up front on the builder to get that more fluent client usage

yep, totally agreed here, and hopefully we can have clean abstractions for future extensions, too, if need be